### PR TITLE
#9285 support for CQL function parsing and XML conversion

### LINKED ISSUE #9265: User Guide - Longitudinal profile plugin
## Description

Create new section for the Longitudinal profile plugin on [User Guide](https://docs.mapstore.geosolutionsgroup.com/en/latest/user-guide/home-page/)

*Documentation section involved*
- [x] User Guide
- [ ] Developer Guide

## Other useful information
Related to #9241

### LINKED ISSUE #9285: Support for functions in CQL filter parser and OGC XML conversion
## Description

Introduce support for parsing cql functions and translating into XML OGC filters. 

## Acceptance criteria

The utils functions should be able
- [x] parse CQL filters like `jsonArrayContains(\"property1\", 'key', 'value') = false AND jsonPointer(\"property2\", 'key') = 'value'`
- [x] translate the CQL filters above in an equivalent XML OGC fitler.

## Other useful information

This feature do not need a UI test. Only unit tests are involved.

### PR DESCRIPTION
**Pull request by offtherailz**
## Description

This PR introduce the capabilities to parse: 

- CQL filter functions
- WKT geometries
- Improves conversion from CQL to OGC XML filter supporting functions and geometries.

This PR allows additional tools like panels injected in the filter panel, to support advanced filtering functionalities. In particular it has been developed to support jsonArrayContains and jsonPointer. 

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [ ] Bugfix
 - [x] Feature
 - [ ] Code style update (formatting, local variables)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [ ] Other... Please describe:

## Issue

**What is the current behavior?**

Actually parsing CQL do not support WKT and functions.

**What is the new behavior?**

Fix #9285, introducing the support for parsing CQL functions, WKT and to convert these into XML OGC filters.

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

## Other useful information

This is an code only change. No new UI test needs to be added.

### CODE REVIEW

File: web/client/utils/ogc/Filter/CQL/parser.js [hunk 1]
```diff
@@ -5,29 +5,31 @@
  * This source code is licensed under the BSD-style license found in the
  * LICENSE file in the root directory of this source tree.
  */
-const fromWKT = () => {
-    throw new Error("WKT parsing for CQL filter not supported yet");
-}; // TODO: use wkt-parser
-const spatialOperators = {
+import {toGeoJSON} from '../../WKT';
```
Comment (MV88) [inline web/client/utils/ogc/Filter/CQL/parser.js:8]: can this be changed to **toGeoJSONGeometry**

File: web/client/utils/ogc/Filter/CQL/parser.js [hunk 2]
```diff
@@ -344,33 +397,43 @@ const buildAst = (tokens) => {
...
-     * @param cqlFilter the cql_filter o parse
-     * @return a javascript representation of the filter.
-     */
-    read: (text) => buildAst(tokenize(text))
-};
+
+
+/**
+ * Parse a CQL filter. returns an AST object representation of the filter.
+ * For the moment this parser doesn't support WKT parsing.
```
Comment (MV88) [inline web/client/utils/ogc/Filter/CQL/parser.js:404]: should this be removed or edited?
```suggestion
 * For the moment this parser doesn't support WKT parsing.
```

File: web/client/utils/ogc/Filter/fromObject.js [hunk 3]
```diff
@@ -26,16 +28,41 @@ const operators = {
...
+        return filterBuilder.propertyName(name);
+    }
     if (includes(logical, type)) {
         return filterBuilder[type](
             ...filters.map(fromObject(filterBuilder))
         );
     }
     if (includes(cql, type)) {
-        return "";
+        return ""; // TODO: implement in filterBuilder as empty filter
```
Comment (MV88) [inline web/client/utils/ogc/Filter/fromObject.js:44]: what about this?

File: web/client/utils/ogc/WKT/toGeoJSON.js [hunk 4]
```diff
@@ -0,0 +1,144 @@
+
```
Comment (MV88) [inline web/client/utils/ogc/WKT/toGeoJSON.js:1]: copyrights

File: web/client/utils/ogc/WKT/__tests__/toGeoJSON-test.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
+import expect from 'expect';
```
Comment (MV88) [inline web/client/utils/ogc/WKT/__tests__/toGeoJSON-test.js:1]: copyrights

File: web/client/utils/ogc/WKT/toGeoJSON.js [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+const parsePoint = (coordinates) => {
```
Comment (MV88) [inline web/client/utils/ogc/WKT/toGeoJSON.js:2]: add jsdoc for these functions

**Review by MV88 — CHANGES_REQUESTED** [review]
LGTM

just a few things

File: web/client/utils/ogc/Filter/fromObject.js (hunk 3 above)
Comment (offtherailz) [reply web/client/utils/ogc/Filter/fromObject.js:44]: Include and Exclude are only part of CQL/ECQL of geoserver. They do not have a proper equivalent in OGC filters.
This is a suggestion for a future implementation of emptyFilter in filterBuilder, equivalent to the "include" cql filter

File: web/client/utils/ogc/Filter/CQL/parser.js (hunk 1 above)
Comment (offtherailz) [reply web/client/utils/ogc/Filter/CQL/parser.js:8]: I think it is a correct name. WKT provides only syntax for geometries, so it is implicit.

Comment (MV88) [reply web/client/utils/ogc/Filter/CQL/parser.js:8]: I strongly disagree and at this point is pointless discuss further so I'm going to approve it as you did.

**Review by MV88 — CHANGES_REQUESTED** [review]
why toGeoJSON is a let and not exported as const?

can this be changed?

**Comment by offtherailz** [comment]
@MV88 `toGeoJSON` is required by the function to transform `parseGeometryCollection` and vice-versa. 
So the function have to be defined before the other. Anyway I don't see the problem. `const` make sense inside the module.
They are not "exported as const" or "exported as let". 

https://stackoverflow.com/questions/32558514/javascript-es6-export-const-vs-export-let

**Review by MV88 — APPROVED** [review]

**Comment by MV88** [comment]
@ElenaGallo please test in DEV

**Comment by offtherailz** [comment]
No additional UI test needed for this PR. Only existing test for the next release.

**Pull request merged**